File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.cpp
class Solution {
public:
    int solve(int ind,int buy,int transactions,vector<int>& prices,vector<vector<vector<int>>>& dp){
        //memoized solotion
        //Base Case we can do anything if the we performed all transcations
        if(ind==prices.size()|| transactions==0){
            return 0;
        }

        if(dp[ind][buy][transactions]!=-1) return dp[ind][buy][transactions];
        
        int Profit=0;

        if(buy){
        int Buy=-prices[ind]+solve(ind+1,0,transactions,prices,dp);
            int Notbuy=0+solve(ind+1,1,transactions,prices,dp);
            Profit=max(Buy,Notbuy);
        }
        else{
            // when we are not allowed to buy we can sell or notSell
            // a transaction would be considered completed if we sell
            int sell=prices[ind]+solve(ind+1,1,transactions-1,prices,dp);
            int notSell=0+solve(ind+1,0,transactions,prices,dp);
            Profit=max(sell,notSell);
        }
            return dp[ind][buy][transactions]=Profit;
    }
    int maxProfit(vector<int>& prices) {

        int n=prices.size();

        vector<vector<vector<int>>> dp(n+1,vector<vector<int>> (2,vector<int>(3,-1)));
        
        return solve(0,1,2,prices,dp);
    }
};

File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.ts

function maxProfit(prices: number[]): number {
    let n:number = prices.length;

    let dp:number[][][][] = Array.from({length:n+1},()=>{
        return Array.from({length:2},()=>{
            return Array.from({length:2},()=>{
                return Array.from({length:3},()=> 0);
            })
        })
    });
    //edge case 1: when count is 2
    for(let i=0;i<n;i++){
        for(let buy=0;buy<2;buy++){
            for(let sell=0;sell<2;sell++){
                dp[i][buy][sell][2]=0;
            }
        }
    }
    //edge case 2:when index is n
    for(let buy=0;buy<2;buy++){
        for(let sell=0;sell<2;sell++){
            for(let count=0;count<3;count++){
                dp[n][buy][sell][count]=0;
            }
        }
    }


    for(let i:number=n-1;i>=0;i--){
        for(let buy:number=0;buy<2;buy++){
            for(let sell:number=0;sell<2;sell++){
                for(let count:number=0;count<3;count++){
                    if(buy===sell) continue;
                    if(buy===1){
                        dp[i][buy][sell][count]=Math.max(-prices[i]+dp[i+1][0][1][count],dp[i+1][1][0][count]);
                    }
                    else{
                        if(count<2){
                            dp[i][buy][sell][count]=Math.max(prices[i]+dp[i+1][1][0][count+1],dp[i+1][0][1][count]);
                        }
                    }
                }
            }
        }
    }
    

    return dp[0][1][0][0];
};